Parse remaining SNR info tables

meew0 opened the pull request:
Primarily, I've added parsing logic for the `offset_72` (`bupmode` data), `offset_76` (`chars` sprites and descriptions), and `offset_80` (`chars` selection grids) info tables. These don't use static structs like the other info tables, but instead variable length "segments" akin to instructions in the main script. I've called them "segments" in the code too based on the terminology I used for kaleido a long time ago; if you'd prefer a different name let me know. Also let me know if you'd like any of the parsing code restructured, this is the first time actually using `binrw` and I might have missed some features that make everything more elegant 😅 

I've also renamed many of the existing `unk*` fields to more descriptive names, and added a bunch of documentation comments where appropriate.

DCNick3 commented:
![image](https://user-images.githubusercontent.com/10363282/221378083-d6ee9a2b-3661-482a-820b-e87f72b84943.png)
Can you pls rebase onto the current master? I did a force-push to remove a failed release (maybe I can find a better workflow for doing this :sweat:), so this branch now references a commit that is no longer in master

DCNick3 commented on shin-core/src/format/scenario/info.rs:
```diff
@@ -6,16 +6,26 @@ use binrw::file_ptr::FilePtrArgs;
 use binrw::{BinRead, BinResult, BinWrite, Endian, FilePtr32};
 use std::io::{Read, Seek};
 
+/// Represents a mask, a black and white image specifying a transition between two screens.
```
For `MaskInfoItem` and other asset references:

I... Don't like how this is worded

The `MaskInfoItem` doesn't _really_ represent a mask, it's merely a reference to it. This is better off in docs for `shin_core::format::mask::MaskTexture` and/or the struct in shin (when it will have masks). Probably not idea to touch other docs in this PR, just a thing to document later

Also, maybe it's worth adding a doc-link for `shin_core` struct representing the asset format and a shin struct representing the loaded asset (like `shin::asset::picture::Picture`)

meew0 replied:
That's fair; I reworded stuff a bit such that it says more “references” instead of “represents”.

As for the doc links to `shin` itself: my thought was that `shin_core` would be useful on its own for people who want to make their own shin-related projects, so I avoided direct references to other parts of the engine. But if you think it would be useful regardless I can still add some links.

DCNick3 replied:
Well, adding a link to `shin_core`'s format impl might still be useful.

As for the `shin` itself...  I am fine with no link being present here. Now that I think about it, I am not sure you can make such a link in rustdoc :thinking:.

DCNick3 commented on shin-core/src/format/scenario/info.rs:
```diff
@@ -39,10 +59,15 @@ impl BustupInfoItem {
+/// Represents the metadata for a background music (BGM) track.
 #[derive(Debug, BinRead, BinWrite)]
 pub struct BgmInfoItem {
+    /// The internal name of the BGM track, corresponding to the filename the engine will look for when playing the BGM.
```
It's actually a name in the `/bgm` directory, excluding the file extension! Ditto for bustup, picture, mask & stuff below.

Smth like "The internal name of the BGM track, corresponding to the filename the engine will look for in the `/bgm` directory when playing the BGM."?

meew0 commented:
Done!

DCNick3 commented on shin-core/src/format/scenario/info.rs:
```diff
@@ -80,34 +115,220 @@ impl MovieInfoItem {
+        character_id: u16,
+
+        /// Defines how the portrait and the corresponding sprite is displayed, and how it behaves on Execute/Resurrect:
+        ///  - **in game**: `0` = empty portrait, `1` = alive, `2` = dead (small blood splotch on sprite), `3` = very dead (large blood splotch on sprite)
```
very dead xD

DCNick3 replied:
Btw, sounds like an ideal case for a enum

meew0 replied:
turns out I was a bit wrong about how exactly this field works, so in the end I didn't end up making an enum here; but I added enums for other relevant fields

DCNick3 commented on shin-core/src/format/scenario/info.rs:
```diff
@@ -80,34 +115,220 @@ impl MovieInfoItem {
+        /// The connector's Y position on the grid.
+        grid_y: u8,
+
+        /// The shape of this connector:
```
Symbols from https://www.compart.com/en/unicode/block/U+2500 might make the docs less ambiguous

DCNick3 commented on shin-core/src/format/scenario/info.rs:
```diff
@@ -80,34 +115,220 @@ impl MovieInfoItem {
+}
+pub type MusicBoxInfo = Vec<MusicBoxInfoItem>;
+
+trait FinalSegment {
```
Also, maybe make a parsing-only helper do find the final segment? As far as I understand, all the `End` variants are signaled by `0` byte. Then you can have a universal type like this:

```rust
#[derive(BinRead, BinWrite)]
enum EndableSegment<T: BinRead + BinWrite> { // TODO: bounds are wrong, don't remember the correct ones
  #[brw(magic = 0x0u8)]
  End,
  Some(T),
}
```

This can be used in `parse_terminated_segment_list` and then discarded, producing just a `Vec` of segments

meew0 replied:
Yes that's definitely a more elegant solution; changed it. (As a consequence the `End` segment no longer shows up in the parsed data, but I don't think that's really an issue)

DCNick3 commented on shin-core/src/format/scenario/info.rs:
```diff
@@ -80,34 +115,220 @@ impl MovieInfoItem {
+}
+
+#[derive(Debug, BinRead, BinWrite)]
+pub struct CharsGridInfoItem {
```
What do the item ids correspond to? Episodes?

meew0 replied:
IDs 0 to 7 correspond to the final grids for episode 1-8 that can be accessed from the title screen. The other IDs (there are like 150) correspond to individual grids representing the state of the characters at specific points of an episode. The currently active grid is selected using the `CHARS` (`0xe0`) instruction

DCNick3 replied:
Now I get it! thx

DCNick3 commented:
Thanks for the docs, I really struggle with writing them consistently :sweat_smile: 

I... think calling them segments is fine... Not that I have a better name idea tbh

meew0 commented on shin-core/src/format/scenario/info.rs:
```diff
@@ -65,12 +92,20 @@ impl SeInfoItem {
-    pub unk3: i16,
+
+    /// The ID of the picture that will be displayed instead of the movie after the movie has finished playing. This is only really relevant for movies used in animations; the movies used in cutscenes have this set to 0.
+    pub linked_picture_id: u16,
```
Done

meew0 commented on shin-core/src/format/scenario/info.rs:
```diff
@@ -80,34 +115,220 @@ impl MovieInfoItem {
+/// An individual instruction for building the data underlying the Character Box (`bupmode`).
+#[derive(Debug, BinRead, BinWrite)]
+pub enum CharacterBoxSegment {
+    /// Defines an individual background to be available for selection in the character box.
```
It's just the backdrop for the posed character. Changed the doc comment to mention this.

meew0 commented on shin-core/src/format/scenario/info.rs:
```diff
@@ -80,34 +115,220 @@ impl MovieInfoItem {
+    #[brw(magic = 0x0u8)]
+    Background {
+        /// The index of the picture that constitutes the primary background image (shown in front).
+        primary_picture_id: u16,
```
Yes; mentioned this here and where else it was relevant (in addition to the module level docs)

meew0 commented:
Addressed the comments. I also did some more testing to find out what the remaining unknown fields meant — the final field of `BgmInfoItem`, and the fields of `CharsSpriteSegment`. Investigating the latter also revealed I was slightly wrong about the `behavior` fields in `CharsGridSegment::Portrait`; those actually just specify defaults that are loaded from the `CharsSpriteSegment`, so I changed that as well.

meew0 commented on shin-core/src/format/scenario/info.rs:
```diff
@@ -80,34 +115,220 @@ impl MovieInfoItem {
+    #[brw(magic = 0x0u8)]
+    End,
+
+    /// Defines a portrait on the grid, with a corresponding character to have its full sprite, name, and description shown when the portrait is selected.
```
On second thought there's a duplicate "shown" in your suggestion 😅 so I changed it slightly

DCNick3 replied:
Whoops :sweat_smile:

DCNick3 commented:
LGTM now

DCNick3 commented:
Thanks!
